### DISCUSSION TITLE
How should I call the functions generated by SDE?

### DISCUSSION THREAD

**Discussion by nicola907** [climateinteractive/SDEverywhere#326] [Q&A]
Hi everyone.
I want to create a videogame with Unreal Engine, that is based on C/C++, which simulates a Vensim model (mymodel.mdl). 
I generated the C/C++ code using the following command:

`sde build mymodel.mdl`

After implementing the missing functions (please see discussion "Errors generating C code from MDL file" for more details) I got a code which compiled correctly. So I took a look at the C/C++ file representing my model and I found the following auto generated functions: initConstants, initLevels, evalLevels, evalAux and other functions that are called by these ones.
I noticed that the variables that I am interested in (my outputs) are calculated in the evalLevels and evalAux functions. Therefore I organized my code in the following manner:

```
//I start out by calling the functions with a name starting with "init..."
//because usually they are the first to be called
initConstants();
initLevels();

//In the following line I set my inputs using a function (setF) that I have defined and implemented
setF(f1, f2, f3);
//Then I try to go on with the simulation by incrementing the auto generated _time variable defined in model.cpp
for (int month = initial_month; month < initial_month + 12; month++) {
	//The setTime function, that I have defined and implemented, modifies the _time variable
	setTime(month);
	evalLevels();//Here I calculate the levels
}
evalAux();//Here I calculate the auxiliaries
//In the following line I get the results by calling a function that I have defined and implemented (getOutputStructure) 
//which returns only the variables that I am interested in
Output output = getOutputStructure();
```

The problem is that the results that I got are not exactly as expected therefore there must be some problems in this sequence of calls...

**Comment by chrispcampbell** [comment]
@nicola907: Alongside the generated code (`mymodel.c` in your case), there is other code in `model.c` that represents the C-level API for interacting with your model.  It already has entry points that run the generated code in the correct order, setting up model input values and reading back output values, so you shouldn't need to manually write control loop code like in your example above.

I would recommend using the `runModelWithBuffers` entry point:

```c
/**
 * Run the model, reading inputs from the given `inputs` buffer, and writing outputs
 * to the given `outputs` buffer.
 *
 * This function performs the same steps as the original `run_model` function,
 * except that it uses the provided pre-allocated buffers.
 *
 * The `inputs` buffer is assumed to have one double value for each input variable;
 * they must be in exactly the same order as the variables are listed in the spec file.
 *
 * After each step of the run, the `outputs` buffer will be updated with the output
 * variables.  The buffer needs to be at least as large as:
 *   `number of output variables` * `number of save points`
 * where `number of save points` is typically one point for each year inclusive of
 * the start and end times.
 *
 * The outputs will be stored in the same order as the outputs are defined in the
 * spec file, with one "row" for each variable.  For example, the first value in
 * the buffer will be the output value at t0 for the first output variable, followed
 * by the output value for that variable at t1, and so on.  After the value for tN
 * (where tN is the last time in the range), the second variable outputs will begin,
 * and so on.
 */
void runModelWithBuffers(double* inputs, double* outputs)
```

There is also a `run_model` entry point which is more geared towards the command line use case (taking in a C input string and writing back out to the console), but I think `runModelWithBuffers` is more appropriate for your use case and it's also the one used by default in the JS `runtime` packages.

You can inspect how `runModelWithBuffers` works by looking around the rest of `model.c` (see the `run` function for example).

**Reply by nicola907** [reply]
Thank you very much for the answer. I will try to use this function.

**Reply by nicola907** [reply]
I have a few questions:
1) Which is the spec file?
2) What is the meaning of SAVEPER?
3) I have noticed that the setInputsFromBuffer functions is not implemented:
`void setInputsFromBuffer(double* inputData) {}`

**Reply by chrispcampbell** [reply]
1. The spec file is a JSON file that allows you to specify what inputs and outputs to include in the generated model.  See [this part](https://github.com/climateinteractive/SDEverywhere#specify-input-and-output-variables) of the README.  You can pass it on the command line like `sde build --spec spec.json <modelname>`.  See `models/arrays_varname` for one example (there are others in the `models` directory).
2. `SAVEPER` is the frequency that output variables are saved, for example, 1 means one data point per time unit (e.g., year), 0.5 means one data point for every half year, etc.
3. Once you provide a spec file that includes some "inputs" (usually these are defined as simple constants in the Vensim model), the `setInputsFromBuffer` will be generated to include the inputs you specify.

**Reply by nicola907** [reply]
Thank you a lot. Now I make the spec.json and see what I get.
